### DISCUSSION TITLE
Any help getting the LSP working with Helix?

### DISCUSSION THREAD

**Discussion by bancomputers** [zk-org/zk#294] [Help] [answered]
If you don't know of it, [Helix](https://helix-editor.com)  is a modal text editor that has native LSP support. I've been having trouble getting the LSP working correctly using the [built in language configurations](https://docs.helix-editor.com/languages.html). It seems to recognize the LSP and allow me to select between four commands (.index, .new, .list, and .tag.list) but none of them seem to do anything. No sign of auto-completion at all either.  The following is my languages.toml file. 

[[language]]
name = "zk"
roots = ["x.md"]
scope = "text.md"
file-types = ["md"]
language-server = { command = "zk", args = ["lsp"] }

I'm not well versed in how LSPs work exactly so any help in getting this to work would be a big help! Thanks!

**Comment by bancomputers** [comment, marked as answer]
I found all my solutions in rather roundabout ways, with a few caveats. I'll list them here in case anyone else might find them useful. 

First, I replaced my 'languages.toml' entry:

    [[language]]
    name = "markdown"
    scope = "source.md"
    injection-regex = "md|markdown|
    file-types = ["md", "markdown]
    roots = [".zk"]
    language-server = { command = "zk", args=["lsp"] }


This is modeled after the default 'languages.toml' included with helix. 

(Psst... add this to the same file to get some sweet syntax highlighting!)

    [[grammar]]
    name = "markdown"
    source = { git = "https://github.com/MDeiml/tree-sitter-markdown", rev = "7e7aa9a25ca9729db9fe22912f8f47bdb403a979", subpath = "tree-sitter-markdown" }

This change immediately fixed my auto-completion and diagnostic issues. As for the others, helix's in-house 'go-to-definition' command works perfectly to follow links (this is bound to 'g' in normal mode followed by 'd'). 

The four commands are still semi-responsive. I can see now that the 'zk.index' command is actually indexing my notebook, as proven by attempting to auto-complete new tags before and after using the command. But, using ':sh zk index' is both quicker and includes a floating results box. The other three commands still don't seem to do much of anything. Using the same ':sh' trick with other commands has mixed results. It works with 'zk list...' using the same floating box as index (though I'm sure this will get very impractical if your list results are numerous). Any other command such as ':sh zk new...' or ':sh zk edit' causes helix to freeze up and forces me to restart, so don't do that! 

Helix's "space mode" (pressing 'space' in normal mode) seems to solve many problems.  Selecting a word and pressing 'space-a' while in normal mode, opens a dialog allowing you to create and link to a new note with the selected word as a title. For a short preview of the first few lines of a file, 'space-k' while the caret is over a link in normal mode works well. For a general command, 'space-f' opens a fuzzy finder for the current folder allowing you to easily search your notebook without crashing helix. Cool!

I hope my misadventures help any other zk users considering this new modal text editor. Helix has been very fun so far and has offered a fairly smooth transition from neovim, even without plugin support yet! 

Happy trails!

**Reply by quesada** [reply]
Hello from 2024!

In case anyone finds this, I tried this config and it didn't work for me. Here's what happened:
https://www.reddit.com/r/HelixEditor/comments/1b2wvtf/help_getting_the_zk_lsp_markdown_with_wikilinks/?

**Reply by sumit-sahrawat** [reply]
@quesada This works okay for me when placed in `.helix/languages.toml` in the root of my notebook. Haven't tested this extensively though.

```toml
[language-server]
zk-lsp = { command = "zk", args = [ "lsp" ] }

[[language]]
name = "markdown"
roots = [ ".zk" ]
language-servers = [ "zk-lsp" ]
```

The syntax highlighting/grammar change is also merged into helix already so it's better to not add that to get the latest version in use by helix.

**Reply by apraga** [reply]
With the configuration above, I have link completion but cannot run `zk.index` or `zk.new` as they requires the database as path:

```
2025-07-20T13:25:38.754 helix_term::commands::lsp [ERROR] execute LSP command: protocol error: 
InvalidRequest: zk.list expects a notebook path as first argument
```

I tried `:lsp-workspace-command zk.nix "."` or with the full path without success : `zk.index /home/alex/notes` is not supported by any langage server.
@sumit-sahrawat Did you manage to get it working ? Thanks !

**Reply by apraga** [reply]
Edit: `spc-a` works to create link from text. The only thing I'm missing to far is to search by title, i.e `zk edit -i`. As my filenames use id rather than slugs, `spc-f` is not very practical. @bancomputers if you have any tips, that would be great !
